txpipe/twopoint_fourier.py

```diff
@@ -348,137 +358,177 @@ def select_calculations(self, nbins_source, nbins_lens):
         Cdd=sacc.standard_types.galaxy_density_cl
 
         type_name = NAMES[k]
-        print(f"Process {self.rank} calcluating {type_name} spectrum for bin pair {i},{i}")
+        print(f"Process {self.rank} calculating {type_name} spectrum for bin pair {i},{j}")
         sys.stdout.flush()
 
-        # We need the theory spectrum for this pair
-        theory = cl_theory[(i,j,k)]
 
         # The binning information - effective (mid) ell values and
         # the window information
         ls = ell_bins.get_effective_ells()
         win = [ell_bins.get_window(b) for b,l  in enumerate(ls)]
 
+        # The healpix pixel windows.  C_ell estimates on healpix
+        # maps (like all pixelized maps) are modulated by a pixel
+        # window function, which we calculate here so we can remove
+        # it below.  These are trivially fast to get, so no point
+        # caching.
+        if (pixel_scheme.name == 'healpix') and self.config['pixel_window']:
+            # Get the raw pixel window functions
+            pixwin_t, pixwin_p = healpy.pixwin(pixel_scheme.nside, True)
+            # Interpolate to our ell values
+            pixwin_ell = np.arange(pixwin_t.size)
+            pixwin_t = np.interp(ls, pixwin_ell, pixwin_t)
+            pixwin_p = np.interp(ls, pixwin_ell, pixwin_p)
+        else:
+            # TODO figure out what these are supposed to be
+            pixwin_t = 1.0
+            pixwin_p = 1.0
+
+        # We need the theory spectrum for this pair
+        #TODO: when we have templates to deproject, use this.
+        theory = cl_theory[(i,j,k)]
+        ell_guess = cl_theory['ell']
+        cl_guess = None
+
         if k == SHEAR_SHEAR:
-            workspace = w22
             field_i = f_wl[i]
             field_j = f_wl[j]
-            cl_guess = [theory, np.zeros_like(theory), np.zeros_like(theory), np.zeros_like(theory)]
-            results_to_use = [(0, CEE), (3, CBB)]
+            results_to_use = [(0, CEE, pixwin_p**2), (3, CBB, pixwin_p**2)]
 
         elif k == POS_POS:
-            workspace = w00
             field_i = f_d[i]
             field_j = f_d[j]
-            cl_guess = [theory]
-            results_to_use = [(0, Cdd)]
+            results_to_use = [(0, Cdd, pixwin_t**2)]
 
         elif k == SHEAR_POS:
-            workspace = w02
             field_i = f_wl[i]
             field_j = f_d[j]
-            cl_guess = [theory, np.zeros_like(theory)]
-            results_to_use = [(0, CdE), (1, CdB)]
+            results_to_use = [(0, CdE, pixwin_t*pixwin_p), (1, CdB, pixwin_t*pixwin_p)]
+
+        if pixel_scheme.name == 'healpix':
+            workspace = nmt.NmtWorkspace()
+        elif pixel_scheme.name == 'gnomonic':
+            workspace = nmt.NmtWorkspaceFlat()
+        else:
+            raise ValueError(f"No NaMaster workspace for pixel scheme {pixel_scheme.name}")
+
+        # Compute mode-coupling matrix
+        workspace.compute_coupling_matrix(field_i, field_j, ell_bins)
```

Comment:
Just a reminder that it's possible to save these to file and read them if you want to avoid having to recalculate them every time.

Reply:
Also, if masks are the same (as may usually be the case for clustering), then you don't need to recompute things.

Reply:
Yep - will definitely implement that once we have multiple lens bins.